File: src/types/note.ts
export type CategoryType = 'Personal' | 'Work' | 'Study' | 'Ideas' | 'Shopping' | string;

export interface Category {
  id: string;
  name: string;
  color: string; // Hex color string
  icon?: string;
  isCustom?: boolean;
}

export type SortBy = 'newest' | 'oldest' | 'a-z' | 'lastEdited';

export type ThemeMode = 'light' | 'dark' | 'system';

export type ViewMode = 'grid' | 'list';

export interface Note {
  id: string;
  title: string;
  description: string;
  category: CategoryType;
  color: string; // Background color for note card in hex
  isPinned: boolean;
  isFavorite: boolean;
  createdAt: number; // Timestamp ms
  updatedAt: number; // Timestamp ms
}

export interface NoteFilter {
  searchQuery: string;
  category: string | null; // null means 'All'
  pinnedOnly: boolean;
  favoritesOnly: boolean;
  sortBy: SortBy;
}

export interface AppSettings {
  themeMode: ThemeMode;
  defaultViewMode: ViewMode;
  hasSeenSplash: boolean;
}

File: flutter_app/lib/models/note.dart
import 'package:hive/hive.dart';

part 'note.g.dart';

@HiveType(typeId: 0)
class Note extends HiveObject {
  @HiveField(0)
  final String id;

  @HiveField(1)
  String title;

  @HiveField(2)
  String description;

  @HiveField(3)
  String category;

  @HiveField(4)
  int colorHex;

  @HiveField(5)
  bool isPinned;

  @HiveField(6)
  bool isFavorite;

  @HiveField(7)
  final DateTime createdAt;

  @HiveField(8)
  DateTime updatedAt;

  Note({
    required this.id,
    required this.title,
    required this.description,
    required this.category,
    required this.colorHex,
    this.isPinned = false,
    this.isFavorite = false,
    required this.createdAt,
    required this.updatedAt,
  });

  Note copyWith({
    String? id,
    String? title,
    String? description,
    String? category,
    int? colorHex,
    bool? isPinned,
    bool? isFavorite,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Note(
      id: id ?? this.id,
      title: title ?? this.title,
      description: description ?? this.description,
      category: category ?? this.category,
      colorHex: colorHex ?? this.colorHex,
      isPinned: isPinned ?? this.isPinned,
      isFavorite: isFavorite ?? this.isFavorite,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}
